enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+
+This alternative registry could be:
+
+- a user-supplied external registry + pull/push credential ( strawman: quay.io )
+- an MCO-supplied minimal registry such that the MCO would be able to serve and store the images it produces.
+- A customer deployed 3rd party registry such as artifactory, etc.
+
+For a secured user-supplied external registry, we would need to make sure that the credentials were available such that: 
+
+- the MCO can push the final image and 
+- rpm-ostree can pull the image during deployment
+
+An external registry would also carry with it extra network overhead, as it could potentially be hosted outside the cluster, but that is an optimization problem to be dealt with later.
+
+An MCO-supplied registry would need to be robust enough to scale to support multiple updating pools at once and be available during a control-plane upgrade, so a simple in-memory registry running on a single node is probably not sufficient.
+
+We do not desire to recreate the internal registry piece-by-piece, so the functionality provided by any MCO-supplied registry would be absolutely minimal and limited only to what would be required to support the MCO's layering operations.
+
+Related Spikes:
+https://issues.redhat.com/browse/MCO-281
+https://issues.redhat.com/browse/MCO-286
+
+#### Image reproducibility
+
+For images we would build in "Phase 1" and "Phase 2", the desire would eventually be that:
+
+- As long as we have the same customer base image and the same MCO content, the output of our image build should always be the same and result in the same image
+
+This would allow for easy change detection, as we'd be able to figure out if an image changed just by observing its digest. It would also make storing and preserving the MCO-built image less important -- we could always rebuild it, the only cost would be time.
+
+Unfortunately, this is not currently a "solved problem" in the container image space, and we as the MCO don't have full control of the problem space to solve it. All it takes is for a single byte to be out of place -- something as simple as a timestamp -- and it's unclear whether we will be able to solve this in the near future.
+
+For now we're assuming we will not have 100% reproducible images and are planning accordingly.
+
+#### Future Upgrade flow
+1. The "base image" will be part of the release image, the same as today.
+1. The CVO will replace a ConfigMap in the MCO namespace with the OS payload reference, as it does today.
+1. The MCO will update an `imagestream` object (e.g. `openshift-machine-config-operator/rhel-coreos`) when this ConfigMap changes.
+1. The MCO builds a container image for each `MachineConfigPool`, using the new base image
+
+1. Each machineconfig pool will support a `custom-external-coreos` imagestream for pulling externally built images
+1.MCD continues to perform drains and reboots,, but for the most part does not change files in `/etc` or install packages per node
```

Comment:
Nit: when rendered this isn't a separate bullet point since its missing a space after the `1.`. Also two commas